core/src/sample_performance_service.rs

```diff
@@ -0,0 +1,100 @@
+use solana_ledger::blockstore::Blockstore;
+use solana_ledger::blockstore_meta::PerfSample;
+use solana_runtime::bank_forks::BankForks;
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, RwLock,
+    },
+    thread::{self, sleep, Builder, JoinHandle},
+    time::{Duration, Instant},
+};
+
+const SAMPLE_INTERVAL: u64 = 60;
+const SLEEP_INTERVAL: u64 = 500;
+
+pub struct SamplePerformanceSnapshot {
+    pub num_transactions: u64,
+    pub num_slots: u64,
+}
+
+pub struct SamplePerformanceService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl SamplePerformanceService {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(
+        bank_forks: &Arc<RwLock<BankForks>>,
+        blockstore: &Arc<Blockstore>,
+        exit: &Arc<AtomicBool>,
+    ) -> Self {
+        let exit = exit.clone();
+        let blockstore = blockstore.clone();
+        let bank_forks = bank_forks.clone();
+
+        info!("Starting SamplePerformance service");
+        let thread_hdl = Builder::new()
+            .name("sample-performance".to_string())
+            .spawn(move || {
+                Self::run(bank_forks, &blockstore, exit);
+            })
+            .unwrap();
+
+        Self { thread_hdl }
+    }
+
+    pub fn run(
+        bank_forks: Arc<RwLock<BankForks>>,
+        blockstore: &Arc<Blockstore>,
+        exit: Arc<AtomicBool>,
+    ) {
+        let forks = bank_forks.read().unwrap();
+        let bank = forks.working_bank();
+        let highest_slot = forks.highest_slot();
+        drop(forks);
+
+        let mut sample_snapshot = SamplePerformanceSnapshot {
+            num_transactions: bank.transaction_count(),
+            num_slots: highest_slot,
+        };
+
+        let mut now = Instant::now();
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                break;
+            }
+
+            let elapsed = now.elapsed();
+
+            if elapsed.as_secs() >= SAMPLE_INTERVAL {
+                now = Instant::now();
+                let bank_forks = bank_forks.read().unwrap();
+                let bank = bank_forks.working_bank();
+                let highest_slot = bank_forks.highest_slot();
+                drop(bank_forks);
+
+                let perf_sample = PerfSample {
+                    num_slots: highest_slot - sample_snapshot.num_slots,
+                    num_transactions: bank.transaction_count() - sample_snapshot.num_transactions,
```

Comment:
potentially this can go negative with forking

Reply:
@sakridge Oh that's interesting. Just the delta for number of transactions in a sample?

Reply:
Yea, because the working bank could build off an older slot.

For example, say you had slot 2 and slot 3 both build from slot 1 which has 1 transaction. slot 2 has 10 transactions, slot 3 has 5 transactions. When you look at slot 2, you will be at 11 transactions total, but then when working bank is at slot 3, your transaction count will only be 6.

Reply:
I think the delta wouldn't be very representative. I was pondering switching the tx count sample to use the root bank. This means the data would be offset by about 12sec. Since that isn't ideal, perhaps we could switch to using the highest available optimistically confirmed/singleGossip once that is exposed (talking to @carllin about that right now).

Reply:
Yea. you need to decide how you are counting transactions... will it be all transactions that landed in all forks since the last sampling, or as you say only rooted transactions. In any case... don't put code that could panic in the validator :)

Reply:
@oJshua Might be helpful to find out how solanabeach is calculating TPS.
Bench_tps just does a simple subtract from working_bank; the analog here would be `bank.transaction_count().checked_sub(sample_snapshot.num_transactions).unwrap_or_default()`

Reply:
@sakridge thanks for bringing this up! @CriesofCarrots I'll see what I can find out.

Reply:
@CriesofCarrots You mentioned using the rooted transactions would be offset by ~12 sec? That might not be so bad.

Reply:
Agreed, not so bad. We could start there and update later if we need more currency.

Reply:
I like that plan.